### ISSUE TITLE
Improve image quality assessment with specialized models

### ISSUE THREAD

**Issue by ajammies** [ajammies/bookbug#79]
## Overview

Currently using Claude Vision for image quality assessment, which is flexible but expensive (~$0.02/image) and slow (~3s). Research suggests an ensemble approach using specialized open-source models could reduce costs by 80%+ while improving accuracy.

## Proposed Ensemble Pipeline

| Stage | Model | Cost | Speed | Purpose |
|-------|-------|------|-------|---------|
| 1. Fast Screen | TOPIQ/Q-Align | Free | 50ms | Catch obviously bad images |
| 2. Artifact Check | HADM | Free | 100ms | Detect hands/face issues |
| 3. Character Match | CLIP-I/DINO | Free | 100ms | Compare to sprite sheet |
| 4. Detailed Analysis | Claude Vision | $0.02 | 3s | Only for borderline cases |

### 1. General Quality Assessment - [IQA-PyTorch](https://github.com/chaofengc/IQA-PyTorch)

```python
pip install pyiqa

import pyiqa
topiq = pyiqa.create_metric('topiq_nr', device='cuda')
qalign = pyiqa.create_metric('qalign', device='cuda')  # ICML 2024, AI-image specific
```

### 2. AI Artifact Detection - [HADM](https://github.com/wangkaihong/HADM)

Specifically trained to detect:
- Distorted/missing/extra body parts
- Wrongly textured faces
- Extra fingers, missing limbs
- Anatomical inconsistencies

### 3. Character Consistency - CLIP/DINO Embeddings

```python
from transformers import CLIPProcessor, CLIPModel
# Compare embeddings of reference sprite vs generated image
similarity = cosine_similarity(ref_embedding, gen_embedding)
```

Metrics:
- **CLIP-I** - Image embedding similarity
- **DINO-I** - Better for fine-grained details

## Implementation Approach

```python
def quality_pipeline(image_path, reference_sprite_path):
    # Stage 1: Fast quality screen
    quality_score = topiq(image_path)
    if quality_score < 0.5:
        return {"pass": False, "reason": "Low quality", "use_llm": False}
    
    # Stage 2: Artifact detection
    artifacts = hadm_detect(image_path)
    if artifacts.has_issues:
        return {"pass": False, "reason": artifacts.issues, "use_llm": False}
    
    # Stage 3: Character consistency
    similarity = clip_similarity(image_path, reference_sprite_path)
    if similarity < 0.7:
        return {"pass": False, "reason": "Character mismatch", "use_llm": False}
    
    # Stage 4: Only use Claude for borderline cases
    if quality_score < 0.7 or similarity < 0.85:
        return claude_detailed_analysis(image_path, context)
    
    return {"pass": True, "score": quality_score}
```

## Research Sources

- [IQA-PyTorch](https://github.com/chaofengc/IQA-PyTorch) - Comprehensive quality assessment toolbox
- [Q-Align](https://github.com/Q-Future/Q-Align) - ICML 2024, unified model for AI image scoring
- [HADM](https://github.com/wangkaihong/HADM) - Human artifact detection dataset & models
- [CLIP-AGIQA](https://arxiv.org/abs/2408.15098) - CLIP-based AI image quality assessment
- [IP-Adapter](https://huggingface.co/docs/diffusers/en/using-diffusers/ip_adapter) - Reference image matching approach

## Tasks

- [ ] Install and test pyiqa with TOPIQ and Q-Align
- [ ] Evaluate HADM for artifact detection
- [ ] Implement CLIP/DINO similarity for character matching
- [ ] Create ensemble pipeline that gates Claude usage
- [ ] Benchmark cost/speed/accuracy vs current approach
- [ ] Consider if custom fine-tuning needed for character consistency

## Notes

- No custom training needed for general quality + artifacts
- Character consistency may benefit from fine-tuning a small classifier on our character designs
- Style adherence works well with CLIP embeddings without training
